crates/re_viewport/src/space_view_entity_picker.rs

```diff
@@ -13,73 +13,49 @@ use crate::{
         compute_heuristic_context_for_entities, is_entity_processed_by_class,
         HeuristicFilterContextPerEntity,
     },
+    ViewportBlueprint,
 };
 
 /// Window for adding/removing entities from a space view.
-#[derive(Clone)]
+///
+/// Delegates to [`re_ui::modal::ModalHandler`]
+#[derive(Default)]
 pub struct SpaceViewEntityPicker {
-    pub space_view_id: SpaceViewId,
+    space_view_id: Option<SpaceViewId>,
+    modal_handler: re_ui::modal::ModalHandler,
 }
 
 impl SpaceViewEntityPicker {
+    pub fn open(&mut self, space_view_id: SpaceViewId) {
+        self.space_view_id = Some(space_view_id);
+        self.modal_handler.open();
+    }
+
     #[allow(clippy::unused_self)]
     pub fn ui(
         &mut self,
+        ui: &mut egui::Ui,
         ctx: &ViewerContext<'_>,
-        ui: &egui::Ui,
-        space_view: &SpaceViewBlueprint,
-    ) -> bool {
-        // This function fakes a modal window, since egui doesn't have them yet: https://github.com/emilk/egui/issues/686
-
-        // In particular, we dim the background and close the window when the user clicks outside it
-        let painter = egui::Painter::new(
-            ui.ctx().clone(),
-            egui::LayerId::new(egui::Order::PanelResizeLine, egui::Id::new("DimLayer")),
-            egui::Rect::EVERYTHING,
-        );
-        painter.add(egui::Shape::rect_filled(
-            ui.ctx().screen_rect(),
-            egui::Rounding::ZERO,
-            egui::Color32::from_black_alpha(128),
-        ));
-
-        // Close window using escape button.
-        let mut open = ui.input(|i| !i.key_pressed(egui::Key::Escape));
-        let title = "Add/remove Entities";
-
-        let response = egui::Window::new(title)
-            // TODO(andreas): Doesn't center properly. `pivot(Align2::CENTER_CENTER)` seems to be broken. Also, should reset every time
-            .default_pos(ui.ctx().screen_rect().center())
-            .collapsible(false)
-            .default_height(640.0)
-            .resizable(true)
-            .frame(egui::Frame {
-                fill: ui.visuals().panel_fill,
-                inner_margin: re_ui::ReUi::view_padding().into(),
-                ..Default::default()
-            })
-            // We do a custom title bar for better adhoc styling.
-            // TODO(andreas): Ideally the default title bar would already adhere to that style
-            .title_bar(false)
-            .show(ui.ctx(), |ui| {
-                title_bar(ctx.re_ui, ui, title, &mut open);
+        viewport_blueprint: &ViewportBlueprint,
+    ) {
+        self.modal_handler.ui(
+            ctx.re_ui,
+            ui,
+            || re_ui::modal::Modal::new("Add/remove Entities"),
+            |_, ui| {
+                let Some(space_view_id) = &self.space_view_id else {
+                    return;
+                };
+
+                let Some(space_view) = viewport_blueprint.space_views.get(space_view_id) else {
+                    return;
+                };
```

Comment:
The modal dialog should probably close itself if it runs into one of these error cases for whatever reason